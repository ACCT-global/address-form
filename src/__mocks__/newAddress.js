export default {
  addressId: '1',
  addressType: 'residential',
  city: null,
  complement: null,
  country: null,
  geoCoordinates: null,
  neighborhood: null,
  number: null,
  postalCode: null,
  receiverName: null,
  reference: null,
  state: null,
  street: null,
}
