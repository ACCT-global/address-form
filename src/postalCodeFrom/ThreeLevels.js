import React, { Component } from 'react'
import PropTypes from 'prop-types'
import AddressShape from '../propTypes/AddressShape'
import find from 'lodash/find'
import map from 'lodash/map'

class ThreeLevels extends Component {
  constructor(props) {
    super(props)

    this.state = getLevels(props.rules)
  }

  componentWillReceiveProps(nextProps) {
    this.setState(getLevels(nextProps.rules))
  }

  handleFirstLevelChange = e => {
    const value = e.target.value

    this.props.onChangeAddress({
      ...this.props.address,
      [this.state.firstLevel.name]: value,
    })
  };

  handleSecondLevelChange = e => {
    const value = e.target.value

    this.props.onChangeAddress({
      ...this.props.address,
      [this.state.secondLevel.name]: value,
    })
  };

  handleThirdLevelChange = e => {
    const value = e.target.value

    this.props.onChangeAddress({
      ...this.props.address,
      ...this.deComposeValue(value),
    })
  };

  composeValue = address =>
    (address[this.state.thirdLevel.name] && address.postalCode
      ? `${address[this.state.thirdLevel.name]}___${address.postalCode}`
      : '');

  deComposeValue = value => {
    const [thirdLevelValue, postalCode] = value.split('___')
    return { [this.state.thirdLevel.name]: thirdLevelValue, postalCode }
  };

  render() {
    const { address, rules } = this.props
    const { firstLevel, secondLevel, thirdLevel } = this.state

    return (
      <div>
        <label>
          {firstLevel.label}
          <select
            name={firstLevel.name}
            value={address[firstLevel.name] || ''}
            onChange={this.handleFirstLevelChange}
          >
            <option value="" />
            {firstLevel.options.map(firstLevelName => (
              <option key={firstLevelName} value={firstLevelName}>
                {firstLevelName}
              </option>
            ))}
          </select>
        </label>
        <label>
          {secondLevel.label}
          <select
            name={secondLevel.name}
            value={address[secondLevel.name] || ''}
            onChange={this.handleSecondLevelChange}
          >
            <option value="" />
            {address[firstLevel.name] &&
              secondLevel.optionsMap[address[firstLevel.name]]
              ? map(
                  secondLevel.optionsMap[address[firstLevel.name]],
                  secondLevelName => (
                    <option key={secondLevelName} value={secondLevelName}>
                      {secondLevelName}
                    </option>
                  )
                )
              : null}
          </select>
        </label>

        <label>
          {thirdLevel.label}
          <select
            name={thirdLevel.name}
            value={this.composeValue(address)}
            onChange={this.handleThirdLevelChange}
          >
            <option value="" />
            {address[firstLevel.name] &&
              address[secondLevel.name] &&
              rules.thirdLevelPostalCodes[address[firstLevel.name]] &&
              rules.thirdLevelPostalCodes[address[firstLevel.name]][
                address[secondLevel.name]
              ]
              ? map(
                  rules.thirdLevelPostalCodes[address[firstLevel.name]][
                    address[secondLevel.name]
                  ],
                  ({ postalCode, thirdLevelName }) => (
                    <option
                      key={thirdLevelName}
                      value={this.composeValue({
                        [thirdLevel.name]: thirdLevelName,
                        postalCode,
                      })}
                    >
                      {thirdLevelName}
                    </option>
                  )
                )
              : null}
          </select>
        </label>
      </div>
    )
  }
}

function getLevels(rules) {
  const firstLevel = find(
    rules.fields,
    ({ name }) => name === rules.postalCodeLevels[0]
  )
  const secondLevel = find(
    rules.fields,
    ({ name }) => name === rules.postalCodeLevels[1]
  )
  const thirdLevel = find(
    rules.fields,
    ({ name }) => name === rules.postalCodeLevels[2]
  )

  return { firstLevel, secondLevel, thirdLevel }
}

ThreeLevels.propTypes = {
  address: PropTypes.shape(AddressShape),
  rules: PropTypes.object.isRequired,
  onChangeAddress: PropTypes.func.isRequired,
}

export default ThreeLevels
