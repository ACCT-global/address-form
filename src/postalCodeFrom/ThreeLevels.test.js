import React from 'react'
import renderer from 'react-test-renderer'
import ThreeLevels from './ThreeLevels'
import { shallow } from 'enzyme'
import BOL from '../country/BOL'
import newAddress from '../__mocks__/newAddress'

describe('ThreeLevels', () => {
  it('without first and second level selected', () => {
    const address = {
      ...newAddress,
      country: 'BOL',
    }

    const tree = renderer
      .create(
        <ThreeLevels
          address={address}
          rules={BOL}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with first level selected and second level not selected', () => {
    const address = {
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
    }

    const tree = renderer
      .create(
        <ThreeLevels
          address={address}
          rules={BOL}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with first and second level selected', () => {
    const address = {
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
      city: 'Méndez',
    }

    const tree = renderer
      .create(
        <ThreeLevels
          address={address}
          rules={BOL}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with first, second and third level selected and not postal code', () => {
    const address = {
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
      city: 'Méndez',
      neighborhood: 'Canasmoro',
    }

    const tree = renderer
      .create(
        <ThreeLevels
          address={address}
          rules={BOL}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('with first, second and third level and postalCode selected', () => {
    const address = {
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
      city: 'Méndez',
      neighborhood: 'Canasmoro',
      postalCode: '90400',
    }

    const tree = renderer
      .create(
        <ThreeLevels
          address={address}
          rules={BOL}
          onChangeAddress={jest.fn()}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should change the first level', () => {
    const handleChange = jest.fn()
    const wrapper = shallow(
      <ThreeLevels
        address={{
          ...newAddress,
          country: 'BOL',
        }}
        rules={BOL}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Tarija' } }
    wrapper.find('select[name="state"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
    })
  })

  it('should change the second level', () => {
    const handleChange = jest.fn()
    const wrapper = shallow(
      <ThreeLevels
        address={{
          ...newAddress,
          country: 'BOL',
          state: 'Tarija',
        }}
        rules={BOL}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Méndez' } }
    wrapper.find('select[name="city"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
      city: 'Méndez',
    })
  })

  it('should change the third level and postal code ', () => {
    const handleChange = jest.fn()
    const wrapper = shallow(
      <ThreeLevels
        address={{
          ...newAddress,
          country: 'BOL',
          state: 'Tarija',
          city: 'Méndez',
        }}
        rules={BOL}
        onChangeAddress={handleChange}
      />
    )

    const event = { target: { value: 'Canasmoro___90400' } }
    wrapper.find('select[name="neighborhood"]').simulate('change', event)

    expect(handleChange).toHaveBeenCalledWith({
      ...newAddress,
      country: 'BOL',
      state: 'Tarija',
      city: 'Méndez',
      neighborhood: 'Canasmoro',
      postalCode: '90400',
    })
  })
})
