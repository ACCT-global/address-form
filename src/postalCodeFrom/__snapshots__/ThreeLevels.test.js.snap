// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThreeLevels with state and city selected 1`] = `
<div>
  <label>
    department
    <select
      name="state"
      onChange={[Function]}
      value="Tarija"
    >
      <option
        value=""
      />
      <option
        value="Beni"
      >
        Beni
      </option>
      <option
        value="Chuquisaca"
      >
        Chuquisaca
      </option>
      <option
        value="Cochabamba"
      >
        Cochabamba
      </option>
      <option
        value="La Paz"
      >
        La Paz
      </option>
      <option
        value="Oruro"
      >
        Oruro
      </option>
      <option
        value="Pando"
      >
        Pando
      </option>
      <option
        value="Potosí"
      >
        Potosí
      </option>
      <option
        value="Santa Cruz"
      >
        Santa Cruz
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    province
    <select
      name="city"
      onChange={[Function]}
      value="Méndez"
    >
      <option
        value=""
      />
      <option
        value="Aniceto Arce"
      >
        Aniceto Arce
      </option>
      <option
        value="Burnet O'Connor"
      >
        Burnet O'Connor
      </option>
      <option
        value="Cercado"
      >
        Cercado
      </option>
      <option
        value="Gran Chaco"
      >
        Gran Chaco
      </option>
      <option
        value="Méndez"
      >
        Méndez
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    city
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Canasmoro___90400"
      >
        Canasmoro
      </option>
      <option
        value="San Lorenzo___90401"
      >
        San Lorenzo
      </option>
      <option
        value="Tomatitas___90402"
      >
        Tomatitas
      </option>
    </select>
  </label>
</div>
`;

exports[`ThreeLevels with state selected and city not selected 1`] = `
<div>
  <label>
    department
    <select
      name="state"
      onChange={[Function]}
      value="Tarija"
    >
      <option
        value=""
      />
      <option
        value="Beni"
      >
        Beni
      </option>
      <option
        value="Chuquisaca"
      >
        Chuquisaca
      </option>
      <option
        value="Cochabamba"
      >
        Cochabamba
      </option>
      <option
        value="La Paz"
      >
        La Paz
      </option>
      <option
        value="Oruro"
      >
        Oruro
      </option>
      <option
        value="Pando"
      >
        Pando
      </option>
      <option
        value="Potosí"
      >
        Potosí
      </option>
      <option
        value="Santa Cruz"
      >
        Santa Cruz
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    province
    <select
      name="city"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Aniceto Arce"
      >
        Aniceto Arce
      </option>
      <option
        value="Burnet O'Connor"
      >
        Burnet O'Connor
      </option>
      <option
        value="Cercado"
      >
        Cercado
      </option>
      <option
        value="Gran Chaco"
      >
        Gran Chaco
      </option>
      <option
        value="Méndez"
      >
        Méndez
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    city
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
    </select>
  </label>
</div>
`;

exports[`ThreeLevels with state, city and neighborhood selected and not postal code 1`] = `
<div>
  <label>
    department
    <select
      name="state"
      onChange={[Function]}
      value="Tarija"
    >
      <option
        value=""
      />
      <option
        value="Beni"
      >
        Beni
      </option>
      <option
        value="Chuquisaca"
      >
        Chuquisaca
      </option>
      <option
        value="Cochabamba"
      >
        Cochabamba
      </option>
      <option
        value="La Paz"
      >
        La Paz
      </option>
      <option
        value="Oruro"
      >
        Oruro
      </option>
      <option
        value="Pando"
      >
        Pando
      </option>
      <option
        value="Potosí"
      >
        Potosí
      </option>
      <option
        value="Santa Cruz"
      >
        Santa Cruz
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    province
    <select
      name="city"
      onChange={[Function]}
      value="Méndez"
    >
      <option
        value=""
      />
      <option
        value="Aniceto Arce"
      >
        Aniceto Arce
      </option>
      <option
        value="Burnet O'Connor"
      >
        Burnet O'Connor
      </option>
      <option
        value="Cercado"
      >
        Cercado
      </option>
      <option
        value="Gran Chaco"
      >
        Gran Chaco
      </option>
      <option
        value="Méndez"
      >
        Méndez
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    city
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Canasmoro___90400"
      >
        Canasmoro
      </option>
      <option
        value="San Lorenzo___90401"
      >
        San Lorenzo
      </option>
      <option
        value="Tomatitas___90402"
      >
        Tomatitas
      </option>
    </select>
  </label>
</div>
`;

exports[`ThreeLevels with state, city, neighborhood and postalCode selected 1`] = `
<div>
  <label>
    department
    <select
      name="state"
      onChange={[Function]}
      value="Tarija"
    >
      <option
        value=""
      />
      <option
        value="Beni"
      >
        Beni
      </option>
      <option
        value="Chuquisaca"
      >
        Chuquisaca
      </option>
      <option
        value="Cochabamba"
      >
        Cochabamba
      </option>
      <option
        value="La Paz"
      >
        La Paz
      </option>
      <option
        value="Oruro"
      >
        Oruro
      </option>
      <option
        value="Pando"
      >
        Pando
      </option>
      <option
        value="Potosí"
      >
        Potosí
      </option>
      <option
        value="Santa Cruz"
      >
        Santa Cruz
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    province
    <select
      name="city"
      onChange={[Function]}
      value="Méndez"
    >
      <option
        value=""
      />
      <option
        value="Aniceto Arce"
      >
        Aniceto Arce
      </option>
      <option
        value="Burnet O'Connor"
      >
        Burnet O'Connor
      </option>
      <option
        value="Cercado"
      >
        Cercado
      </option>
      <option
        value="Gran Chaco"
      >
        Gran Chaco
      </option>
      <option
        value="Méndez"
      >
        Méndez
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    city
    <select
      name="neighborhood"
      onChange={[Function]}
      value="Canasmoro___90400"
    >
      <option
        value=""
      />
      <option
        value="Canasmoro___90400"
      >
        Canasmoro
      </option>
      <option
        value="San Lorenzo___90401"
      >
        San Lorenzo
      </option>
      <option
        value="Tomatitas___90402"
      >
        Tomatitas
      </option>
    </select>
  </label>
</div>
`;

exports[`ThreeLevels without state and city selected 1`] = `
<div>
  <label>
    department
    <select
      name="state"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
      <option
        value="Beni"
      >
        Beni
      </option>
      <option
        value="Chuquisaca"
      >
        Chuquisaca
      </option>
      <option
        value="Cochabamba"
      >
        Cochabamba
      </option>
      <option
        value="La Paz"
      >
        La Paz
      </option>
      <option
        value="Oruro"
      >
        Oruro
      </option>
      <option
        value="Pando"
      >
        Pando
      </option>
      <option
        value="Potosí"
      >
        Potosí
      </option>
      <option
        value="Santa Cruz"
      >
        Santa Cruz
      </option>
      <option
        value="Tarija"
      >
        Tarija
      </option>
    </select>
  </label>
  <label>
    province
    <select
      name="city"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
    </select>
  </label>
  <label>
    city
    <select
      name="neighborhood"
      onChange={[Function]}
      value=""
    >
      <option
        value=""
      />
    </select>
  </label>
</div>
`;
